File: apps/api/app/workflow_readiness.py
from __future__ import annotations

from app.models import SpaceRecord, WorkflowReadiness


def build_workflow_readiness(space: SpaceRecord) -> WorkflowReadiness:
    total_objects = len(space.objects)
    approved_count = sum(1 for item in space.objects if item.status == "Approved")
    reviewed_count = sum(1 for item in space.objects if item.status == "Reviewed")
    pending_review_count = sum(1 for item in space.objects if item.status == "Needs Review")

    blockers: list[str] = []
    if total_objects == 0:
        blockers.append("no-objects")
    if pending_review_count > 0:
        blockers.append("pending-review")
    if approved_count == 0:
        blockers.append("no-approved")

    export_ready = total_objects > 0 and pending_review_count == 0
    publication_ready = export_ready and approved_count > 0

    return WorkflowReadiness(
        approvedCount=approved_count,
        blockers=blockers,
        exportReady=export_ready,
        listingReady=publication_ready,
        pendingReviewCount=pending_review_count,
        reviewedCount=reviewed_count,
        shareReady=publication_ready,
        storyReady=publication_ready,
        totalObjects=total_objects,
    )

File: apps/web/src/lib/workflow-readiness.ts
import type { ObjectRecord, SpaceRecord } from "@/lib/platform-types"

export type WorkflowBlockerCode = "no-objects" | "pending-review" | "no-approved-objects"

export type RoomChecklistItem = {
  id: string
  name: string
  objectCount: number
  pendingReviewCount: number
  ready: boolean
}

export type WorkflowReadiness = {
  approvedCount: number
  blockers: WorkflowBlockerCode[]
  exportReady: boolean
  pendingReviewCount: number
  publishReady: boolean
  readyCount: number
  reviewedCount: number
  roomChecklist: RoomChecklistItem[]
  shareReady: boolean
  totalObjects: number
}

export function getWorkflowReadiness(
  space: Pick<SpaceRecord, "rooms" | "objects">,
  objectOverride?: ObjectRecord[],
): WorkflowReadiness {
  const objects = objectOverride ?? space.objects
  const totalObjects = objects.length
  const approvedCount = objects.filter((objectRecord) => objectRecord.status === "Approved").length
  const reviewedCount = objects.filter((objectRecord) => objectRecord.status === "Reviewed").length
  const pendingReviewCount = objects.filter((objectRecord) => objectRecord.status === "Needs Review").length
  const readyCount = approvedCount + reviewedCount

  const roomChecklist = space.rooms.map((room) => {
    const roomObjects = objects.filter(
      (objectRecord) =>
        objectRecord.roomId === room.id ||
        objectRecord.roomName?.toLowerCase() === room.name.toLowerCase(),
    )
    const roomPendingReviewCount = roomObjects.filter(
      (objectRecord) => objectRecord.status === "Needs Review",
    ).length

    return {
      id: room.id,
      name: room.name,
      objectCount: roomObjects.length,
      pendingReviewCount: roomPendingReviewCount,
      ready: roomObjects.length > 0 && roomPendingReviewCount === 0,
    }
  })

  const blockers: WorkflowBlockerCode[] = []
  if (totalObjects === 0) {
    blockers.push("no-objects")
  }
  if (pendingReviewCount > 0) {
    blockers.push("pending-review")
  }
  if (approvedCount === 0) {
    blockers.push("no-approved-objects")
  }

  const exportReady = totalObjects > 0 && pendingReviewCount === 0
  const publishReady = exportReady && approvedCount > 0
  const shareReady = publishReady

  return {
    approvedCount,
    blockers,
    exportReady,
    pendingReviewCount,
    publishReady,
    readyCount,
    reviewedCount,
    roomChecklist,
    shareReady,
    totalObjects,
  }
}
